README.md
# flutter_refresh
A Flutter plugin for refreshing every scrollable view by pulling down-up.



# Show cases


![](https://github.com/jzoom/flutter_refresh/raw/master/example/res/1.gif)


# Roadmap

>>see:[ROADMAP.md](https://github.com/jzoom/flutter_refresh/blob/master/ROADMAP.md)

# Changelogs

>>see:[CHANGELOG.md](https://github.com/jzoom/flutter_refresh/blob/master/README.md)

# Quick Start


## Installation


1 Add 

```bash

flutter_refresh : ^0.0.1

```
to your pubspec.yaml ,and run 

```bash
flutter packages get 
```
in your project's root directory.




2 Add 

```

import 'package:flutter_refresh/flutter_refresh.dart';

```

and write the code like this:

```


  Future<Null> onFooterRefresh() {
    return new Future.delayed(new Duration(seconds: 2), () {
      setState(() {
        _itemCount += 10;
      });
    });
  }

  
  Future<Null> onHeaderRefresh() {
    return new Future.delayed(new Duration(seconds: 2), () {
      setState(() {
        _itemCount = 10;
      });
    });
  }


  @override
  Widget build(BuildContext context) {
    ...
    return new Refresh(
          onFooterRefresh: onFooterRefresh,
          onHeaderRefresh: onHeaderRefresh,
          childBuilder: (BuildContext context,
              {ScrollController controller, ScrollPhysics physics}) {
            return new Container(
                child: new ListView.builder(
              physics: physics,
              controller: controller,
              itemBuilder: _itemBuilder,
              itemCount: _itemCount,
            ));
          },
        );
  }



```

>> full example see here: [main.dart](https://github.com/jzoom/flutter_refresh/blob/master/example/lib/main.dart).










CHANGELOG.md
## [0.0.1] - [2018/05/23]

    * Basic
    * Support header refresh
    * Support footer refresh

## [0.0.2] - [2018/05/24]
    
    * Format code


test/flutter_refresh_test.dart


import 'package:flutter_refresh/flutter_refresh.dart';
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return new MaterialApp(
        title: 'Flutter Demo',
        theme: new ThemeData(
          primarySwatch: Colors.blue,
        ),
        home: new MyHomePage(title: 'Pull to refresh'),
        localizationsDelegates: [
          //   Refresh.delegate()
        ],
        routes: {
          'route':(BuildContext context){
            return null;
          },
        }
    );
  }
}


class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => new _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> with TickerProviderStateMixin {
  Widget _itemBuilder(BuildContext context, int index) {
    return new Padding(
      key: new Key(index.toString()),
      padding: new EdgeInsets.all(10.0),
      child: new Text(
        "顺序数据:$index ",
        style: new TextStyle(fontSize: 14.0),
      ),
    );
  }

  int _itemCount;

  @override
  void initState() {
    _itemCount = 10;
    super.initState();
  }

  Future<Null> onFooterRefresh() {
    return new Future.delayed(new Duration(seconds: 2), () {
      setState(() {
        _itemCount += 10;
      });
    });
  }

  Future<Null> onHeaderRefresh() {
    return new Future.delayed(new Duration(seconds: 2), () {
      setState(() {
        _itemCount = 10;
      });
    });
  }

  @override
  Widget build(BuildContext context) {



    return new Scaffold(
        appBar: new AppBar(
          title: new Text(widget.title),
        ),
        body: new SafeArea(
            child: new Refresh(
              onFooterRefresh: onFooterRefresh,
              onHeaderRefresh: onHeaderRefresh,
              childBuilder: (BuildContext context,
                  {ScrollController controller, ScrollPhysics physics}) {
                return new Container(
                    child: new ListView.builder(
                      physics: physics,
                      controller: controller,
                      itemBuilder: _itemBuilder,
                      itemCount: _itemCount,
                    ));
              },
            )));
  }
}


void main() {
  testWidgets('adds one to input values', (WidgetTester tester) async {
    await tester.pumpWidget(new MyApp());
  });
}


lib/src/refresh.dart
import 'package:flutter/widgets.dart';
import 'dart:async';
import 'dart:io';
import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';
import 'package:flutter_refresh/src/refresh_child.dart';
import 'package:flutter_refresh/src/refresh_controller.dart';
import 'package:flutter_refresh/src/refresh_widget.dart';

///Only support 3 types of direct child
bool checkChild(Widget src) {
  switch (src.runtimeType) {
    case GridView:
    case SingleChildScrollView:
    case ListView:
      {
        return true;
      }
  }

  return false;
}


typedef Widget RefreshScrollViewBuilder(BuildContext context,
    {ScrollController controller, ScrollPhysics physics});

class Refresh extends StatefulWidget {
  final RefresherCallback onHeaderRefresh;
  final RefresherCallback onFooterRefresh;
  final RefreshController controller;
  final ScrollController scrollController;
  final RefreshScrollViewBuilder childBuilder;
  final ScrollPhysics physics;
  final Widget child;

  static ScrollPhysics createScrollPhysics(ScrollPhysics src) {
    if (Platform.isAndroid) {
      ScrollPhysics physics = new AlwaysScrollableScrollPhysics()
          .applyTo(new BouncingScrollPhysics());
      if (src != null) {
        return physics.applyTo(src);
      }
      return physics;
    }
    return src;
  }

  Refresh(
      {Key key,
      this.childBuilder,
      this.scrollController,
      this.onHeaderRefresh,
      this.child,
      this.controller,
      this.physics,
      this.onFooterRefresh})
      : super(key: key) {
    if (child != null) {
      assert(
          checkChild(child),
          "child must be GridView,SingleChildScrollView,ListView,"
          "if you want to put scrollview into another container, please use Refresh.builder instead");
    }

    if (childBuilder == null && child == null) {
      throw new Exception("childBuild or child must not be both null");
    }
  }
//
//  factory Refresh.singleChild(
//      {Key key, Widget child, RefreshScrollViewBuilder childBuilder}) {
//    assert(
//        checkChild(child),
//        "child must be GridView,SingleChildScrollView,ListView,"
//        "if you want to put scrollview into another container, please use Refresh.builder instead");
//
//    return new Refresh(
//      key: key,
//      childBuilder: childBuilder,
//    );
//  }

  @override
  State<StatefulWidget> createState() {
    return new _RefreshState();
  }
}

typedef void StateHandler(ScrollNotification notification);

class _RefreshState extends State<Refresh> with TickerProviderStateMixin {
  double _headerRefreshOffset = 50.0;
  double _footerRefreshOffset = 50.0;

  RefreshWidgetController _headerValue;
  RefreshWidgetController _footerValue;

  _RefreshHandler _headerHandler;
  _RefreshHandler _footerHandler;

  ScrollController controller;
  StateHandler _state;

  bool _isAnimation = false;
  bool _animationComplete = false;

  void _scrollToAnimationFirst(double offset) {
    if (!_isAnimation) {
      if (_animationComplete) {
        controller.jumpTo(offset);
      } else {
        _isAnimation = true;
        controller
            .animateTo(offset,
                duration: new Duration(milliseconds: 300), curve: Curves.ease)
            .whenComplete(() {
          _isAnimation = false;
          _animationComplete = true;
        });
      }
    }
  }

  void _loading(ScrollNotification notification) {
    if (notification is ScrollUpdateNotification) {
      assert(_hander != null);
      _scrollToAnimationFirst(_hander.getScrollOffset(notification.metrics));
    }
  }

  ///
  ///  State machine:
  ///
  /// Darg(User drag screen)
  /// =>Ready
  /// =>Loading(User end drag and  the condition is ok)
  /// =>Cancel(User end drag and the condition is not ok)
  /// =>End ( complete the action)
  ///
  /// The State convertion :
  ///
  ///   1 End=>Drag            when the state is End and user begin to drag(notification is ScrollStartNotification)
  ///   2 Drag=>Loading/Cancel    Two loading: head or foot,  condition: when user end drag
  ///   3 Loading=>Cancels         The callback has been called
  ///   4 Cancel =>End
  ///
  ///
  bool _handleScrollNotification(ScrollNotification notification) {
    _state(notification);
    return false;
  }
//
//
//  void _cancel(ScrollNotification notification){
//    if (notification is ScrollStartNotification) {
//      _state = _end;
//    } else if (notification is ScrollUpdateNotification) {
//      if(_hander!=null){
//        _hander.cancel(notification.metrics);
//      }
//    }
//  }

  void _end(ScrollNotification notification) {
    if (notification is ScrollStartNotification) {
      if (notification.dragDetails != null) {
        // print("drag start");
        _state = _drag;
        _animationComplete = false;
      }
    } else if (notification is ScrollUpdateNotification) {
      ScrollMetrics metrics = notification.metrics;
      double pixels = metrics.pixels;
      if (pixels < 0) {
        if (_headerValue != null) {
          _headerValue.value = -pixels;
        }
      } else {
        //
        if (_footerValue != null) {
          double extValue = pixels - metrics.maxScrollExtent;
          _footerValue.value = extValue;
        }
      }
    }
  }

  _RefreshHandler _hander;

  void _drag(ScrollNotification notification) {
    if (notification is ScrollUpdateNotification) {
      ScrollMetrics metrics = notification.metrics;
      if (notification.dragDetails != null) {
        double moveValue;
        if (_headerHandler != null &&
            (moveValue = _headerHandler.getRefreshWidgetMoveValue(metrics)) >
                0) {
          _hander = _headerHandler;
        } else if (_footerHandler != null &&
            (moveValue = _footerHandler.getRefreshWidgetMoveValue(metrics)) >
                0) {
          _hander = _footerHandler;
        } else {
          _hander = null;
        }
        if (_hander != null) {
          if (_hander.isReady(moveValue)) {
            _hander.changeState(RefreshState.ready);
          } else {
            _hander.changeState(RefreshState.drag);
          }
          _hander.controller.value = moveValue;
        }
      } else {
        if (_hander != null &&
            _hander.isReady(_hander.getRefreshWidgetMoveValue(metrics))) {
          _hander.loading(metrics).whenComplete(() {
            //loading ok
            assert(_hander != null);

            _hander = null;
            _state = _end;
          });
          _state = _loading;
          return;
        }
        _state = _end;
      }
    }
  }

  @override
  void initState() {
    super.initState();

    _state = _end;
  }

  ScrollController _tryGetController(Widget src) {
    switch (src.runtimeType) {
      case GridView:
      case ListView:
        {
          return (src as BoxScrollView).controller;
        }

      case SingleChildScrollView:
        return (src as SingleChildScrollView).controller;
    }

    return null;
  }

  void _updateState() {
    if (widget.onHeaderRefresh != null) {
      if (_headerValue == null) _headerValue = new RefreshWidgetController();
      if (_headerHandler == null ||
          _headerHandler.callback != widget.onHeaderRefresh) {
        _headerHandler = new _RefreshHeaderHandler(
            controller: _headerValue,
            callback: widget.onHeaderRefresh,
            offset: _headerRefreshOffset);
      }
    } else {
      if (_headerValue != null) _headerValue.dispose();
      _headerHandler = null;
    }

    if (widget.onFooterRefresh != null) {
      if (_footerValue == null) _footerValue = new RefreshWidgetController();
      if (_footerHandler == null ||
          _footerHandler.callback != widget.onFooterRefresh) {
        _footerHandler = new _RefreshFooterHandler(
            controller: _footerValue,
            callback: widget.onFooterRefresh,
            offset: _footerRefreshOffset);
      }
    } else {
      if (_footerValue != null) _footerValue.dispose();
      _footerHandler = null;
    }

    ScrollController controller = widget.scrollController;
    if (controller == null) {
      if (widget.child != null) {
        controller = _tryGetController(widget.child);
      }
    }

    if (controller == null) {
      if (this.controller == null) {
        this.controller = new ScrollController();
      } else {
        //不动
      }
    } else {
      //不等于,是否要移除什么
      this.controller = controller;
    }
  }

  @override
  void didChangeDependencies() {
    _updateState();
    super.didChangeDependencies();
  }

  @override
  void didUpdateWidget(Refresh oldWidget) {
    _updateState();
    super.didUpdateWidget(oldWidget);
  }

  Widget _cloneChild(Widget src) {
    switch (src.runtimeType) {
      case GridView:
        {
          GridView listView = src as GridView;
          return new GridView.custom(
            gridDelegate: listView.gridDelegate,
            childrenDelegate: listView.childrenDelegate,
            controller: controller,
            physics: Refresh.createScrollPhysics(listView.physics),
            key: listView.key,
            scrollDirection: listView.scrollDirection,
          );
        }
        break;
      case SingleChildScrollView:
        {
          SingleChildScrollView listView = src as SingleChildScrollView;
          return new SingleChildScrollView(
            controller: controller,
            physics: Refresh.createScrollPhysics(listView.physics),
            key: listView.key,
            scrollDirection: listView.scrollDirection,
          );
        }
        break;
      case ListView:
        {
          ListView listView = src as ListView;
          return new ListView.custom(
            childrenDelegate: listView.childrenDelegate,
            controller: controller,
            physics: Refresh.createScrollPhysics(listView.physics),
            key: listView.key,
            scrollDirection: listView.scrollDirection,
          );
        }
    }

    return null;
  }

  @override
  Widget build(BuildContext context) {
    Widget notificationChild = new NotificationListener(
      child: widget.child == null
          ? widget.childBuilder(context,
              controller: controller,
              physics: Refresh.createScrollPhysics(widget.physics))
          : _cloneChild(widget.child),
      onNotification: _handleScrollNotification,
    );

    List<Widget> children = [notificationChild];

    if (widget.onHeaderRefresh != null) {
      children.add(new RefreshWidget(
        height: _headerRefreshOffset,
        controller: _headerValue,
        createTween: createTweenForHeader,
        alignment: Alignment.topCenter,
        maxOffset: 300.0,
        childBuilder:
            (BuildContext context, RefreshWidgetController controller) {
          return new DefaultRefreshChild(
            controller: controller,
            icon: new Icon(Icons.arrow_downward),
            up: true,
          );
        },
      ));
    }

    if (widget.onFooterRefresh != null) {
      children.add(new RefreshWidget(
        height: _footerRefreshOffset,
        controller: _footerValue,
        createTween: createTweenForFooter,
        maxOffset: 300.0,
        alignment: Alignment.bottomCenter,
        childBuilder:
            (BuildContext context, RefreshWidgetController controller) {
          return new DefaultRefreshChild(
            controller: controller,
            icon: new Icon(Icons.arrow_upward),
            showLastUpdate: false,
            up: false,
          );
        },
      ));
    }

    return new Stack(
      key: widget.key,
      children: children,
    );
  }
}

abstract class _RefreshHandler {
  final RefreshWidgetController controller;
  final RefresherCallback callback;
  final double offset;

  RefreshState _state = RefreshState.drag;

  _RefreshHandler({this.controller, this.callback, this.offset});

  double getScrollOffset(ScrollMetrics metrics);

  double getRefreshWidgetMoveValue(ScrollMetrics metrics);

  Future<Null> loading(ScrollMetrics metrics) {
    changeState(RefreshState.loading);
    dynamic result = callback();
    assert(
        result is Future,
        "In this version,the call back must return a Future.value(null),  "
        "\n If the app is doing some working in the closure, the closure must define like this : "
        " \n Future<Null> onHeaderRefresh()  async {\n"
        "  await network.doSomeWork(); return new Future.value(null);"
        "\n } ");
    {
      result.whenComplete(() {
        changeState(RefreshState.drag);
      });
    }
    return result;
  }

  bool isReady(double moveValue) {
    return moveValue > offset;
  }

  void changeState(RefreshState currentState) {
    if (_state != currentState) {
      //通知状态改变
      _state = currentState;

      controller.state = _state;
    }
  }

  void cancel(ScrollMetrics metrics);
}

class _RefreshFooterHandler extends _RefreshHandler {
  _RefreshFooterHandler(
      {RefreshWidgetController controller,
      RefresherCallback callback,
      double offset: 50.0})
      : super(controller: controller, callback: callback, offset: offset);

  //When loading start, what value is ScrollView pixils?
  @override
  double getScrollOffset(ScrollMetrics metrics) {
    return metrics.maxScrollExtent + offset;
  }

  @override
  double getRefreshWidgetMoveValue(ScrollMetrics metrics) {
    return metrics.pixels - metrics.maxScrollExtent;
  }

  @override
  void cancel(ScrollMetrics metrics) {
    controller.value = metrics.pixels;
  }
}

class _RefreshHeaderHandler extends _RefreshHandler {
  _RefreshHeaderHandler(
      {RefreshWidgetController controller,
      RefresherCallback callback,
      double offset: 50.0})
      : super(controller: controller, callback: callback, offset: offset);

  //When loading start, what value is ScrollView pixils?
  @override
  double getScrollOffset(ScrollMetrics metrics) {
    return -offset;
  }

  @override
  void cancel(ScrollMetrics metrics) {
    controller.value = metrics.pixels;
  }

  @override
  double getRefreshWidgetMoveValue(ScrollMetrics metrics) {
    return -metrics.pixels;
  }
}


example/lib/flat_list.dart
import 'package:flutter/widgets.dart';

import 'package:flutter_refresh/flutter_refresh.dart';

typedef Widget IndexedDataBuilder(BuildContext context,
    {dynamic item, int index, dynamic extraData});

typedef Widget HeaderFooterBuilder(BuildContext context,
    {List<dynamic> data, dynamic extraData});

class FlatList extends StatelessWidget {
  final HeaderFooterBuilder headerBuilder;
  final HeaderFooterBuilder footerBuilder;
  final IndexedDataBuilder itemBuilder;
  final List<dynamic> data;
  final dynamic extraData;

  //Grid View
  final int numColumns = 0;

  /**
   *
   */
  @override
  Widget build(BuildContext context) {
    int itemCount = data.length;
    int offset = 0;

    if (headerBuilder != null) {
      offset++;
      ++itemCount;
    }

    if (footerBuilder != null) {
      ++itemCount;
    }

    BoxScrollView scrollView;

    if (numColumns > 0) {
      return new GridView.count(
        crossAxisCount: numColumns,
      );
    }

    return new ListView.builder(
      itemBuilder: (BuildContext context, int index) {
        if (index == 0) {
          if (headerBuilder != null) {
            return headerBuilder(context, data: data, extraData: extraData);
          }
        }

        if (index == itemCount - 1) {
          if (footerBuilder != null) {
            return footerBuilder(context, data: data, extraData: extraData);
            ;
          }
        }

        return itemBuilder(context,
            index: index,
            item: this.data[index + offset],
            extraData: extraData);
      },
      itemCount: itemCount,
    );
  }
}


lib/src/refresh_child.dart
import 'refresh_widget.dart';
import 'package:flutter/material.dart';

class DefaultRefreshLocal {
  static DefaultRefreshLocal en(
          {String loading: "loading...",
          String pullDownToRefresh: "pull down to refresh",
          String pullUpToRefresh: "pull up to refresh",
          String releaseToRefresh: "release to refresh",
          String lastUpdate = "last update"}) =>
      new DefaultRefreshLocal(
          lastUpdate: lastUpdate,
          loading: loading,
          releaseToRefresh: releaseToRefresh,
          pullDownToRefresh: pullDownToRefresh,
          pullUpToRefresh: pullUpToRefresh);

  static DefaultRefreshLocal zh(
          {String loading: "加载中...",
          String pullDownToRefresh: "下拉刷新",
          String pullUpToRefresh: "上拉加载更多",
          String releaseToRefresh: "放开加载",
          String lastUpdate = "最后更新"}) =>
      new DefaultRefreshLocal(
          lastUpdate: lastUpdate,
          loading: loading,
          releaseToRefresh: releaseToRefresh,
          pullDownToRefresh: pullDownToRefresh,
          pullUpToRefresh: pullUpToRefresh);

  final String loading;
  final String pullDownToRefresh;
  final String pullUpToRefresh;
  final String releaseToRefresh;
  final String lastUpdate;

  const DefaultRefreshLocal(
      {this.lastUpdate,
      this.loading,
      this.pullUpToRefresh,
      this.pullDownToRefresh,
      this.releaseToRefresh});
}

typedef RefreshChild RefreshChildBuilder(
    BuildContext context, RefreshWidgetController controller);

abstract class RefreshChild extends StatefulWidget {
  final RefreshWidgetController controller;

  RefreshChild({this.controller});

  @override
  State<StatefulWidget> createState();
}

class DefaultRefreshChild extends RefreshChild {
  final bool showState;
  final bool showLastUpdate;
  final Widget icon;
  final DefaultRefreshLocal local;
  final bool up;

  DefaultRefreshChild({
    RefreshWidgetController controller,
    this.showState: true,
    this.showLastUpdate: true,
    this.icon,
    DefaultRefreshLocal local,
    this.up: true,
  })  : this.local = local == null ? DefaultRefreshLocal.zh() : local,
        super(controller: controller);

  @override
  State<StatefulWidget> createState() {
    return new _DefaultRefreshHeaderState();
  }
}

class _DefaultRefreshHeaderState extends State<DefaultRefreshChild>
    with TickerProviderStateMixin {
  AnimationController _animation;

  Tween<double> _tween;

  DateTime _lastUpdate;

  RefreshState _state = RefreshState.drag;

  String _formateDate() {
    DateTime time = _lastUpdate;
    return " ${_twoDigist(time.hour)}:${_twoDigist(time.minute)}";
  }

  static String _twoDigist(int num) {
    if (num < 10) return "0$num";
    return num.toString();
  }

  @override
  void initState() {
    super.initState();

    _animation = new AnimationController(vsync: this);
    _tween = new Tween<double>(
      begin: 0.0,
      end: 1.0,
    );
    _tween.animate(_animation);
    _lastUpdate = new DateTime.now();
  }

  void _rotate(double value, bool releaseToRefresh) {
    _animation
        .animateTo(value,
            duration: new Duration(milliseconds: 200), curve: Curves.ease)
        .whenComplete(() {});
  }

  void _updateState() {
    switch (widget.controller.state) {
      case RefreshState.ready:
        {
          _rotate(0.5, true);
        }

        break;

      case RefreshState.drag:
        {
          _rotate(0.0, true);
        }

        break;
      case RefreshState.loading:
        {}
        break;
    }

    setState(() {
      _state = widget.controller.state;
    });
  }

  @override
  void didChangeDependencies() {
    widget.controller.addStateListener(_updateState);
    super.didChangeDependencies();
  }

  @override
  void didUpdateWidget(RefreshChild oldWidget) {
    if (widget.controller != oldWidget.controller) {
      oldWidget.controller.removeStateListener(_updateState);
      widget.controller.addStateListener(_updateState);
    }
    super.didUpdateWidget(oldWidget);
  }

  @override
  void dispose() {
    widget.controller.removeStateListener(_updateState);
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    TextStyle style =
        Theme.of(context).textTheme.body1; // //new TextStyle(fontSize: 14.0);

    List<Widget> texts = [];
    if (widget.showState) {
      texts.add(
        new Text(
          _state == RefreshState.loading
              ? widget.local.loading
              : (_state == RefreshState.ready
                  ? widget.local.releaseToRefresh
                  : (widget.up
                      ? widget.local.pullDownToRefresh
                      : widget.local.pullUpToRefresh)),
          style: style,
        ),
      );
    }

    if (widget.showLastUpdate) {
      texts.add(new Text(
        "${widget.local.lastUpdate} ${_formateDate()}",
        style: style,
      ));
    }

    Widget text = texts.length > 0
        ? new Padding(
            padding: new EdgeInsets.only(left: 10.0),
            child: new Column(
              mainAxisSize: MainAxisSize.min,
              children: texts,
            ),
          )
        : null;

    List<Widget> row = [
      _state == RefreshState.loading
          ? new SizedBox(
              width: 20.0, height: 20.0, child: new CircularProgressIndicator())
          : new RotationTransition(
              turns: _animation,
              child: widget.icon,
            ),
    ];

    if (text != null) {
      row.add(text);
    }

    return new Row(
      mainAxisSize: MainAxisSize.min,
      children: row,
    );
  }
}


lib/src/refresh_widget.dart
import 'package:flutter/material.dart';
import 'dart:async';

import 'package:flutter_refresh/src/refresh_child.dart';

typedef Future<Null> RefresherCallback();

enum RefreshState {
  //User is dragging the ScrollView,but the offset is not enough
  drag,
  //User is dragging the ScrollView,and the offset is enough
  ready,
  //User has released dragging
  loading,
}

///refresh componet
class RefreshWidgetController extends ValueNotifier<double> {
  RefreshWidgetController()
      : _state = new ValueNotifier(RefreshState.drag),
        super(0.0);

  final ValueNotifier<RefreshState> _state;

  bool get loading => _state.value == RefreshState.loading;

  set state(RefreshState state) {
    _state.value = state;
  }

  RefreshState get state => _state.value;

  @override
  double get value => super.value;

  @override
  set value(double newValue) {
    if (loading) return;
    super.value = newValue;
  }

  void addStateListener(VoidCallback updateState) {
    _state.addListener(updateState);
  }

  void removeStateListener(VoidCallback updateState) {
    _state.removeListener(updateState);
  }

  @override
  void dispose() {
    _state.dispose();
    super.dispose();
  }
}

typedef RectTween CreateTween(RefreshWidget widget);

RectTween createTweenForHeader(RefreshWidget widget) {
  return new RectTween(
    begin: new Rect.fromLTRB(0.0, -widget.height, 0.0, 0.0),
    end: new Rect.fromLTRB(0.0, 300.0, 0.0, 0.0),
  );
}

RectTween createTweenForFooter(RefreshWidget widget) {
  return new RectTween(
    begin: new Rect.fromLTRB(0.0, 0.0, 0.0, widget.height),
    end: new Rect.fromLTRB(0.0, 0.0, 0.0, -300.0),
  );
}

class RefreshWidget extends StatefulWidget {
  final double height;

  final double maxOffset;

  final RefreshChildBuilder childBuilder;

  final RefreshWidgetController controller;

  final CreateTween createTween;

  final AlignmentGeometry alignment;

  RefreshWidget(
      {this.height,
      this.controller,
      this.childBuilder,
      this.createTween,
      this.alignment,
      this.maxOffset})
      : assert(controller != null);

  @override
  State<StatefulWidget> createState() {
    return new _RefreshHeaderState();
  }
}

class _RefreshHeaderState extends State<RefreshWidget>
    with TickerProviderStateMixin {
  //The animation controller to control the widtet's position
  AnimationController _positionController;
  //
  Animation<Rect> _positionFactor;

  @override
  void initState() {
    super.initState();
    _positionController = new AnimationController(vsync: this);
    _positionFactor = widget.createTween(widget).animate(_positionController);
  }

  ///End=>Loading=>End
  void _updateValue() {
    double value = widget.controller.value / (widget.maxOffset + widget.height);
    //let's move head
    _positionController.value = value;
  }

  void _updateState() {
    switch (widget.controller.state) {
      case RefreshState.drag:
        break;
      case RefreshState.loading:
        {
          double value = widget.height / (widget.maxOffset + widget.height);
          _positionController
              .animateTo(value,
                  duration: new Duration(milliseconds: 300), curve: Curves.ease)
              .whenComplete(() {});
        }
        break;
      case RefreshState.ready:
        break;
    }
  }

  @override
  void didChangeDependencies() {
    widget.controller.addListener(_updateValue);
    widget.controller.addStateListener(_updateState);
    super.didChangeDependencies();
  }

  @override
  void didUpdateWidget(RefreshWidget oldWidget) {
    if (oldWidget.controller != widget.controller) {
      oldWidget.controller.removeListener(_updateValue);
      oldWidget.controller.removeStateListener(_updateState);
      widget.controller.addListener(_updateValue);
      widget.controller.addStateListener(_updateState);
    }
    super.didUpdateWidget(oldWidget);
  }

  @override
  void dispose() {
    widget.controller.removeListener(_updateValue);
    widget.controller.removeStateListener(_updateState);
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    final RefreshWidget widget = this.widget;
    return new RelativePositionedTransition(
        size: new Size(0.0, 0.0),
        rect: _positionFactor,
        child: new AnimatedBuilder(
            animation: _positionController,
            builder: (BuildContext context, Widget child) {
              return new Align(
                child: new SizedBox(
                  height: widget.height,
                  child: this.widget.childBuilder(context, widget.controller),
                ),
                alignment: widget.alignment,
              );
            }));
  }
}


lib/src/refresh_controller.dart
import 'package:flutter/foundation.dart';

class RefreshController extends ValueNotifier<int> {
  static const int HEAD_START_REFRESH = 4;
  static const int HEAD_END_REFRESH = 1;
  static const int FOOT_START_REFRESH = 2;
  static const int FOOT_END_REFRESH = 3;

  RefreshController() : super(0);

  void startHeadRefresh() {
    _ensureListeners();
    value = HEAD_START_REFRESH;
  }

  void _ensureListeners() {
    assert(
        hasListeners,
        "The RefresherController must attach to a widget first, "
        "the message is showing probably becourse the method is called before build . ");
  }

  void startFootRefresh() {
    _ensureListeners();
    value = HEAD_END_REFRESH;
  }

  void endHeadRefresh() {
    _ensureListeners();
    value = HEAD_START_REFRESH;
  }

  void endFootRefresh() {
    _ensureListeners();
    value = FOOT_END_REFRESH;
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
/**
 *
 */
import 'package:flutter/foundation.dart';
import 'dart:async';
import 'package:flutter_refresh/flutter_refresh.dart';

void main() => runApp(new MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return new MaterialApp(
      title: 'Flutter Demo',
      theme: new ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: new MyHomePage(title: 'Pull to refresh'),
      localizationsDelegates: [
        //   Refresh.delegate()
      ],
      routes: {
        'route':(BuildContext context){
          return null;
        },
      }
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => new _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> with TickerProviderStateMixin {
  Widget _itemBuilder(BuildContext context, int index) {
    return new Padding(
      key: new Key(index.toString()),
      padding: new EdgeInsets.all(10.0),
      child: new Text("顺序数据:$index",
        style: new TextStyle(fontSize: 14.0),
      ),
    );
  }

  int _itemCount;

  @override
  void initState() {
    Navigator.popUntil(context, (Route<dynamic> route){
      if(route.settings.name == ''){

      }
    });
    /*
    Navigator.of(context).push(new MaterialPageRoute(builder: (BuildContext context){

      return null;

    })).then((data){



    });*/



    _itemCount = 10;
    super.initState();
  }

  Future<Null> onFooterRefresh() {
    return new Future.delayed(new Duration(seconds: 2), () {
      setState(() {
        _itemCount += 10;
      });
    });
  }

  Future<Null> onHeaderRefresh() {
    return new Future.delayed(new Duration(seconds: 2), () {
      setState(() {
        _itemCount = 10;
      });
    });
  }

  @override
  Widget build(BuildContext context) {



    return new Scaffold(
        appBar: new AppBar(
          title: new Text(widget.title),
        ),
        body: new SafeArea(
            child: new Refresh(
          onFooterRefresh: onFooterRefresh,
          onHeaderRefresh: onHeaderRefresh,
          childBuilder: (BuildContext context,
              {ScrollController controller, ScrollPhysics physics}) {
            return new Container(
                child: new ListView.builder(
              physics: physics,
              controller: controller,
              itemBuilder: _itemBuilder,
              itemCount: _itemCount,
            ));
          },
        )));
  }
}
